data/shell-completions/zsh/meson-comp-helper.py

```diff
@@ -0,0 +1,225 @@
+        ))
+
+    def user_first(d):
+        dd = d.copy()
+        try:
+            u = dd.pop('user')
+            yield ('user', u)
+        except KeyError:
+            pass
+        yield from dd.items()
+
+    alt_decls = []
+    alt_words = []
+    functions_per_type = []
+    for k, v in user_first(by_section):
+        tag_prefix = zsh_tag(k)
+        specs_arr = '__ndescgroup_{}'.format(tag_prefix)
+        # No need to quote; `v` already consists of quoted _describe specs.
+        sv = '  local -a {}=(\n    {}\n  )'.format(specs_arr, '\n    '.join(v))
+        alt_decls.append(sv)
+        word = '{}-options:{} option:(( "${{(@){}}}" ))'.format(
+            tag_prefix, k, specs_arr)
+        alt_words.append(quote(word))
+    sv = '\n'.join((
+         '__meson_project_buildoptions() {',
+         '  # This function loops over tag labels itself.',
+         '\n'.join(alt_decls),
+         '  local -a alternative_argv=( "$@" )',
+         '  _alternative -O alternative_argv {}'.format(' '.join(alt_words)),
+         '}\n'))
+    functions_per_type.append(sv)
+    sv = '\n'.join((
+        '__meson_project_buildoption_choices() {',
+        '  # This function loops over tag labels itself.',
+        '  local m_optname="$1"',
+        '  local descr',
+        '  case "$m_optname" in',
+        '\n'.join(switch_entries.values()),
+        '  *)',
+        '    _message "unknown option; cannot offer choices"'
+        '  ;;',
+        '  esac',
+        '}\n',
+    ))
+    functions_per_type.append(sv)
+    funcs = '\n'.join(functions_per_type)
+    return funcs
+
+
+def emit_message_on_exception(exc):
+    from traceback import format_exception_only
+    lines = format_exception_only(type(exc), exc)
+    mfmt = 'cannot complete: {}'
+    msgs = [quote(mfmt.format(line.rstrip())) for line in lines]
+    print('__meson_project_buildoptions () {',
+          '\n'.join('  ' + '_message ' + msg for msg in msgs),
+          '}',
+          '__meson_project_buildoption_choices () {',
+          '\n'.join('  ' + '_message ' + msg for msg in msgs),
+          '}',
+          sep='\n')
+
+
+def path_from_builddir(builddir, mode):
+    return os.path.join(builddir, 'meson-info', f'intro-{mode}.json')
+
+
+def _getopts(argv):
+    builddir = None
+    istream = None
+    mode = None
+    posopts = 0
+    for i in argv:
+        if i.startswith('--'):
+            mode = i[2:]
+        else:
+            if posopts > 0:
+                continue
+            posopts += 1
+            builddir = i
+    if mode is None:
+        raise ValueError("completion mode is not specified")
+    if mode not in ('buildoptions', ):
+        raise ValueError("unknown completion mode: {}".format(mode))
+    if builddir is None or builddir == '-':
+        istream = sys.stdin
+    else:
+        p = path_from_builddir(builddir, mode)
+        istream = open(p)
+    return {
+        'istream': istream,
+        'mode': mode,
+        'builddir': builddir,
+    }
```

Comment:
Oh please no, please use `argparse`

Reply:
An `argparse`-based parser would definitely be much easier to maintain if this ever gets featureful — e. g. if we will have to complete multiple intro information types here. This makes even more sense if we follow @eli-schwartz's suggestion and promote this to an internal script.

The only disadvantage of using `argparse` here I can think of is latency considerations. The script is executed from scratch (i. e. python initialization, etc.) when a shell user presses <kbd>Tab</kbd> after e. g. `meson configure build -D`, so it should not take too long to finish.

I'm leaning on the side of trying it out and maybe revisiting the decision later if latency is a problem.

Reply:
According to [the following experiment](https://gist.github.com/bonktree/3e6acda5cfc48519e7b318655435da57) on a [popular project](https://github.com/pipewire/pipewire) with a lot of user options, the `argparse` overhead is visible, but looks negligible on the scale of total-time-to-response. The bulkiest parts of that seem to reside in either python initialization or the python-shell glue and are unavoidable.

`cProfile`'s profiling data hints at string operations, but that's still a teardrop in a cup.